### PR TITLE
[ALLUXIO-2182]  Added a metrics master for aggregating the client metrics

### PR DESCRIPTION
**Pull request by yupeng9**
https://alluxio.atlassian.net/browse/ALLUXIO-2182

Each client sends over the metrics to the master in its heartbeat. And the new metrics master will aggregate the metrics defined by the aggregators.

This PR adds a client-master heartbeat to send the client-side metrics info. More test coverage improvement will be added once we agree on the architecture.

### CODE REVIEW

**Comment by yupeng9** [comment]
#1 We could add a shutdown hook for the last heartbeat. However, I feel this is a bit overengineered, as the metrics, imo, might not necessary to be exact but more best efforts. Losing some metrics in 1-2s might not be too problematic.
#2 We can have it as a configurable value, but before we see some real use cases or requests, I feel it is not a must-have feature to be in this framework PR.
#3 We could. But I want to ramp up metrics in future PRs, and this one is to set up the foundation.

**Comment by aaudiber** [comment]
1. If the client runs for shorter than the heartbeat interval, it will send no metrics. I think that could be problematic, especially if we need to increase the heartbeat interval to reduce pressure on the master.
2. It seems like a very low-hanging fruit that's much more useful than per-jvm randomly generated ids. We can do it in a different PR if you prefer, but the ID used to identify clients seems like it fits in this PR as a part of the foundation/framework PR for client metrics.
3. Ok, future PR is fine

**Comment by yupeng9** [comment]
@aaudiber 
hmm, for 1, I'm not that concerned, because the number of metrics that can happen within 1-2s should not be large. But I'm okay with adding it if you insist.

For 2, I prefer a separate PR for adding this optional client id.

**Comment by calvinjia** [comment]
@yupeng9 Have you verified there are threadpool issues when running with a large number of clients and does increasing the threadpool solve the issue without any side effects? I think it is important to have a solution which works for larger clusters (as opposed to just turning it off).

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Task that carries the client metrics information to master through heartheat. This class manages
+ * its own {@link MetricsMasterClient}.
+ *
+ * If the task fails to heartbeat to the master, it will destroy its old master client and recreate
+ * it before retrying.
+ */
+@NotThreadSafe
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: Don't we access this from multiple threads despite it being not thread safe?

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  private long mLastSuccessfulHeartbeatMs;
+
+  /**
+   * Constructs a new {@link ClientMasterSync}.
+   *
+   * @param masterClient the master client
+   */
+  public ClientMasterSync(MetricsMasterClient masterClient) {
+    mMasterClient = Preconditions.checkNotNull(masterClient, "masterClient");
+    mHeartbeatTimeoutMs = (int) Configuration.getMs(PropertyKey.WORKER_BLOCK_HEARTBEAT_TIMEOUT_MS);
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: This timeout value seems unrelated?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 3]
```diff
@@ -184,6 +216,13 @@ public synchronized void reset() throws IOException {
     init(MasterInquireClient.Factory.create());
   }
 
+  /**
+   * @return the unique id of the context
+   */
+  public String getId() {
+    return mId;
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: Is there any way to determine the ID of a process, how will users be able to map an ID to a process after seeing the metrics be abnormal?

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      // An error occurred, log and ignore it or error if heartbeat timeout is reached
+      LOG.error("Failed to receive or execute master heartbeat command: {}", e);
+      mMasterClient.disconnect();
+      if (mHeartbeatTimeoutMs > 0) {
+        if (System.currentTimeMillis() - mLastSuccessfulHeartbeatMs >= mHeartbeatTimeoutMs) {
+          if (Configuration.getBoolean(PropertyKey.TEST_MODE)) {
+            throw new RuntimeException("Master heartbeat timeout exceeded: " + mHeartbeatTimeoutMs);
+          }
+          LOG.error("Master heartbeat timeout exceeded: " + mHeartbeatTimeoutMs);
+          System.exit(-1);
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: Do we really want to system exit if the metrics heartbeat times out?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 5]
```diff
@@ -161,6 +190,8 @@ public void close() throws IOException {
     mFileSystemMasterClientPool = null;
     mBlockMasterClientPool.close();
     mBlockMasterClientPool = null;
+    mMetricsMasterClient.close();
+    mMetricsMasterClient = null;
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: What do we want to achieve here, since `close()` will probably cause the heartbeat thread to fail, should we interrupt the heartbeat thread first?

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+   * The method the worker should periodically execute to heartbeat back to the master.
+   *
+   * @param metrics a list of client metrics
+   */
+  public synchronized void heartbeat(final List<Metric> metrics) throws IOException {
+    retryRPC(new RpcCallable<Void>() {
+      @Override
+      public Void call() throws TException {
+        mClient.metricsHeartbeat(FileSystemContext.INSTANCE.getId(),
+            NetworkAddressUtils.getClientHostName(), new MetricsHeartbeatTOptions(metrics));
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: The client id is present in each metric as well as the `metricsHeartbeat` call?

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    mClient = new MetricsMasterClientService.Client(mProtocol);
+  }
+
+  /**
+   * The method the worker should periodically execute to heartbeat back to the master.
+   *
+   * @param metrics a list of client metrics
+   */
+  public synchronized void heartbeat(final List<Metric> metrics) throws IOException {
+    retryRPC(new RpcCallable<Void>() {
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: The nested retries (`retryRPC` and `heartbeatThread`) may cause unneeded complexity?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 8]
```diff
@@ -928,6 +928,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.MASTER)
           .build();
+  public static final PropertyKey MASTER_METRICS_HEARTBEAT_TIMEOUT_MS =
+      new Builder(Name.MASTER_METRICS_HEARTBEAT_TIMEOUT_MS)
+          .setAlias(new String[]{"alluxio.master.metrics.heartbeat.timeout.ms"})
+          .setDefaultValue("5min")
+          .setDescription("The timeout value of metrics master's heartbeats.")
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
+          .setScope(Scope.WORKER)
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Is this the right scope?

File: core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java [hunk 9]
```diff
@@ -40,6 +40,7 @@
   public static final String WORKER_FILESYSTEM_MASTER_SYNC = "Worker FileSystemMaster Sync";
   public static final String WORKER_PIN_LIST_SYNC = "Worker Pin List Sync";
   public static final String WORKER_SPACE_RESERVER = "Worker Space Reserver";
+  public static final String METRICS_MASTER_SYNC = "Metrics Master Sync";
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java]: Perhaps name this similar to the others `Master ...`?

File: core/common/src/main/java/alluxio/metrics/ClientMetrics.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+package alluxio.metrics;
+
+/**
+ * Metrics of an Alluxio client.
+ */
+public final class ClientMetrics {
+  /**
+   * Total number of bytes short-circuit read from local storage.
+   */
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/ClientMetrics.java]: `/** ... */` format

File: core/common/src/thrift/common.thrift [hunk 11]
```diff
@@ -27,8 +27,9 @@ struct BlockLocation {
 struct Metric {
   1: string instance
   2: string hostname
-  3: string name
-  4: i64 value
+  3: string instanceId
+  4: string name
+  5: i64 value
```
Comment (calvinjia) [inline core/common/src/thrift/common.thrift]: This change is backward incompatible?

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java [hunk 12]
```diff
@@ -125,7 +132,20 @@ public void start(Boolean isLeader) throws IOException {
   }
 
   @Override
-  public void putWorkerMetrics(InstanceType instance, String hostname, List<Metric> metrics) {
-    mMetricsStore.putWorkerMetrics(instance, hostname, metrics);
+  public void clientHeartbeat(String clientId, String hostname, List<Metric> metrics) {
+    if (metrics.isEmpty()) {
+      return;
+    }
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: Could we do this logic in `putClientMetrics`, since it should be able to handle an empty set?

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java [hunk 13]
```diff
@@ -22,13 +21,25 @@
...
  */
 public interface MetricsMaster extends Master  {
+  /**
+   * Handles the client's heartbeat request for metrics collection.
+   *
+   * @param clientId the client id
+   * @param hostname the client hostname
+   * @param metrics client-side metrics
+   */
+  void clientHeartbeat(String clientId, String hostname, List<Metric> metrics);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java]: Since `clientId` is in `metrics`, we could make this one API. If we keep multiple APIs, could we name them symmetrically?

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 1 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: This is a heartbeat executor, and it's only invoked once per JVM?

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 2 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: done

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 14]
```diff
@@ -133,6 +149,9 @@ public static FileSystemContext create(Subject subject, MasterInquireClient mast
    */
   private FileSystemContext(Subject subject) {
     mParentSubject = subject;
+    mExecutorService = Executors.newFixedThreadPool(1,
+        ThreadFactoryUtils.build("metrics-master-heartbeat-%d", true));
+    mId = UUID.randomUUID().toString().replace("-", "");
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 3 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: I considered this, but there does not seem a good approach commonly agreed on: https://stackoverflow.com/questions/35842/how-can-a-java-program-get-its-own-process-id?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

We could log the ID after it's created.

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 4 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: Removed and made it best effort to send the metrics info.

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java (hunk 6 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: What do you mean? The client id is carried in the metric sent over to the metrics master as an identifier.

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 8 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/PropertyKey.java]: good catch

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java (hunk 7 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: I tried to be consistent with other heartbeat invocations. For example, in `BlockMasterClient`, we also retry the heartbeat RPC. But I'm also okay not to have retry since the heartbeat would retry anyway in the next beat.

File: core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java (hunk 9 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java]: done

File: core/common/src/main/java/alluxio/metrics/ClientMetrics.java (hunk 10 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/metrics/ClientMetrics.java]: done

File: core/common/src/thrift/common.thrift (hunk 11 above)
Comment (yupeng9) [reply core/common/src/thrift/common.thrift]: Metric is added in this branch, so it's not public yet

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java (hunk 13 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java]: What do you mean? This API only handles heartbeat from the client, while the worker heartbeats are handled in the `BlockMaster`.

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 1 above)
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: Don't we also access it in the shutdown hook?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 3 above)
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: Logging sounds good for now.

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java (hunk 6 above)
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: The client id is in the RPC and in each of the metrics objects, is it necessary?

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java (hunk 7 above)
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: I'm fine with addressing this later in a general refactor, could you open a ticket?

File: core/common/src/thrift/common.thrift (hunk 11 above)
Comment (calvinjia) [reply core/common/src/thrift/common.thrift]: Got it, thanks

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 1 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: That's true.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java (hunk 13 above)
Comment (calvinjia) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java]: I meant the `clientHeartbeat` and `putWorkerMetrics` methods

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java (hunk 6 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: I see. It's needed so when no metric is sent, the master still knows which client their heartbeat is from.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java (hunk 13 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsMaster.java]: Got it.

File: core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java (hunk 7 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/MetricsMasterClient.java]: Created at ALLUXIO-3198. https://alluxio.atlassian.net/browse/ALLUXIO-3198

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 15]
```diff
@@ -147,6 +168,18 @@ private synchronized void init(MasterInquireClient masterInquireClient) {
         new FileSystemMasterClientPool(mParentSubject, mMasterInquireClient);
     mBlockMasterClientPool = new BlockMasterClientPool(mParentSubject, mMasterInquireClient);
     mClosed.set(false);
+    // setup metrics master client sync
+    mMetricsMasterClient = new MetricsMasterClient(MasterClientConfig.defaults());
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: Here, I think you will have to propagate the master inquire client as well as the subject.

`MasterClientConfig.defaults().withSubject(mParentSubject).withMasterInquireClient(mMasterInquireClient)`

Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: That's good to know. Thanks!

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by calvinjia** [comment]
@yupeng9 Could you clarify how the name of the metric gets created? ie. there is some code in `MetricsSystem` which puts it into `instance.hostname.metric` form, but does that get changed later with the introduction of client id?

File: core/client/fs/src/main/java/alluxio/client/block/stream/LocalFilePacketReader.java [hunk 16]
```diff
@@ -69,6 +71,7 @@ public DataBuffer readPacket() throws IOException {
     ByteBuffer buffer = mReader.read(mPos, Math.min(mPacketSize, mEnd - mPos));
     DataBuffer dataBuffer = new DataByteBuffer(buffer, buffer.remaining());
     mPos += dataBuffer.getLength();
+    MetricsSystem.clientCounter(ClientMetrics.BYTES_READ_LOCAL).inc(dataBuffer.getLength());
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/block/stream/LocalFilePacketReader.java]: I think that's different: the metric here collects the short-circuit read, but not the read from remote.
Also, I don't like the API to be modified for metrics collection.

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 17]
```diff
@@ -98,10 +115,22 @@
   /** The parent user associated with the {@link FileSystemContext}. */
   private final Subject mParentSubject;
 
+  /**
+   * @return the instance of file system context
+   */
+  public static FileSystemContext get() {
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: `INSTANCE` also works for most cases. This method is a bit special as we want to defer its instantation for better unit testing.

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 18]
```diff
@@ -98,10 +115,22 @@
...
+    if (sInstance == null) {
+      synchronized (FileSystemContext.class) {
+        sInstance = create();
+      }
+    }
+    return sInstance;
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: Good catch

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 19]
```diff
@@ -133,6 +162,10 @@ public static FileSystemContext create(Subject subject, MasterInquireClient mast
    */
   private FileSystemContext(Subject subject) {
     mParentSubject = subject;
+    mExecutorService = Executors.newFixedThreadPool(1,
+        ThreadFactoryUtils.build("metrics-master-heartbeat-%d", true));
+    mId = IdUtils.createFileSystemContextId();
+    LOG.info("Created filesystem context with id " + mId);
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    mHeartbeatTimeoutMs =
+        (int) Configuration.getMs(PropertyKey.MASTER_METRICS_HEARTBEAT_TIMEOUT_MS);
+    mLastSuccessfulHeartbeatMs = System.currentTimeMillis();
+  }
+
+  @Override
+  public void heartbeat() throws InterruptedException {
+    List<alluxio.thrift.Metric> metrics = new ArrayList<>();
+    for (Metric metric : MetricsSystem.allClientMetrics()) {
+      metric.setInstanceId(FileSystemContext.get().getId());
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: Why not? For now there is only one context per JVM.

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * Constructs a new {@link ClientMasterSync}.
+   *
+   * @param masterClient the master client
+   */
+  public ClientMasterSync(MetricsMasterClient masterClient) {
+    mMasterClient = Preconditions.checkNotNull(masterClient, "masterClient");
+    mHeartbeatTimeoutMs =
+        (int) Configuration.getMs(PropertyKey.MASTER_METRICS_HEARTBEAT_TIMEOUT_MS);
+    mLastSuccessfulHeartbeatMs = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: Yes? This is instantiated at the client.

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 22]
```diff
@@ -418,6 +418,7 @@ public void setPermission(Path path, FsPermission permission) throws IOException
    *      org.apache.hadoop.conf.Configuration)
    */
   //@Override This doesn't exist in Hadoop 1.x, so cannot put {@literal @Override}.
+  @Override
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Right. This is automatically added by the IDE.

**Comment by yupeng9** [comment]
In the client's JVM, the metric's name does not have the client id, but when it is sent to the master, the client-id is added to identify the client.

File: core/common/src/main/java/alluxio/util/IdUtils.java [hunk 23]
```diff
@@ -91,4 +91,11 @@ public static long createSessionId() {
   public static long createMountId() {
     return getRandomNonNegativeLong();
   }
+
+  /**
+   * @return a random UUID which does not have hyphens
+   */
+  public static String createFileSystemContextId() {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/util/IdUtils.java]: could this use a long instead? UUID is very long for something that shows up in metrics graphing UIs, and with the hostname included we only need the ID to be unique among clients on the same host. Even with 100k clients there's less than a 1 in a billion chance of collision.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java [hunk 24]
```diff
@@ -41,15 +45,25 @@ public Object getFieldValue(Metric o) {
...
+  private static String getHostnameAndId(String hostname, String id) {
+    String str = hostname == null ? "" : hostname;
+    str = str.replace('.', '_');
+    str += id == null ? "" : ":" + id;
+    return str;
+  }
 
   private final IndexedSet<Metric> mWorkerMetrics =
-      new IndexedSet<>(FULL_NAME_INDEX, NAME_INDEX, HOSTNAME_INDEX);
+      new IndexedSet<>(FULL_NAME_INDEX, NAME_INDEX, HOSTNAME_ID_INDEX);
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: this is confusing since worker metrics don't include ID. Can we just have two different indexes, one for hostname, one for hostname+id?

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: When using the Hadoop API, we create a different context (`AbstractFileSystem#updateFileSystemAndContext`)

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: We set the last successful heartbeat without actually doing a heartbeat. Since this is in the constructor, could there be an issue where `run` is not called immediately after the constructor causing the first heartbeat to incorrectly believe it has timed out?

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java [hunk 25]
```diff
@@ -60,28 +74,50 @@ public Object getFieldValue(Metric o) {
...
       List<Metric> metrics) {
-    IndexedSet<Metric> set = getMetricsByInstanceType(instance);
-    set.removeByField(HOSTNAME_INDEX, hostname);
+    if (metrics.isEmpty()) {
+      return;
+    }
+    mClientMetrics.removeByField(HOSTNAME_ID_INDEX, getHostnameAndId(hostname, clientId));
     for (Metric metric : metrics) {
       if (metric.getHostname() == null) {
         continue; // ignore metrics who hostname is null
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: why would the hostname be null? Is this check needed for `putWorkerMetrics` as well?

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java [hunk 26]
```diff
@@ -60,28 +74,50 @@ public Object getFieldValue(Metric o) {
...
       }
-      set.add(metric);
+      LOG.info(metric.toString());
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: forgot to remove this?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 27]
```diff
@@ -133,6 +164,13 @@ public static FileSystemContext create(Subject subject, MasterInquireClient mast
    */
   private FileSystemContext(Subject subject) {
     mParentSubject = subject;
+    mExecutorService = Executors.newFixedThreadPool(1,
+        ThreadFactoryUtils.build("metrics-master-heartbeat-%d", true));
+    mId = IdUtils.createFileSystemContextId();
+    LOG.info("Created filesystem context with id {}."
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: We don't have to. This ID  can be used for other purposes, too.

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: I see. Considering @aaudiber's point, this last successful heartbeat time is not that useful. I'll get rid of it.

File: core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java]: I see. Good to know that

File: core/common/src/main/java/alluxio/util/IdUtils.java (hunk 23 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/util/IdUtils.java]: This UI is not displayed graph UI , but it's used for debugging only.
For now this ID is used to identifying a JVM on a node.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 24 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: hmm I feel it's fine as hostname only is a special form of hostname+id.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 25 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: This is a safe guard. It can happen in our simulated local cluster in integration testing, where the same JVM has the master, worker and client metrics.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 26 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: Good catch

File: core/common/src/main/java/alluxio/util/IdUtils.java (hunk 23 above)
Comment (aaudiber) [reply core/common/src/main/java/alluxio/util/IdUtils.java]: I mean in graphana or wherever users choose to consume our metrics from.

If the ID is per-JVM per-node, that makes it even less necessary to have a UUID

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 24 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: this feels like an unnecessary hack to save a few lines of boilerplate code. The data type <hostname> is different from the type <hostname, id>, not a special case of it.

Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: or perhaps we could get rid of the concept of hostname, and just have a concept of "ID". Then the ID for workers would be their hostname, while the ID for clients would be hostname + a client specific ID

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 25 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: should we use it in `putWorkerMetrics` then?

Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: yup, updated `putWorkerMetrics `

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 24 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: having the `ID` sounds good to me.
